io.typefox.yang.parent/io.typefox.yang/src/test/java/io/typefox/yang/tests/YangValidatorTest.xtend

```diff
@@ -0,0 +1,634 @@
+package io.typefox.yang.tests
+
+import io.typefox.yang.yang.Import
+import io.typefox.yang.yang.YangVersion
+import org.junit.Test
+
+import static io.typefox.yang.validation.IssueCodes.*
+
+/**
+ * Validation test for the YANG language.
+ * 
+ * @author akos.kitta
+ */
+class YangValidatorTest extends AbstractYangTest {
+
+	@Test
+	def void checkYangVersion() {
+		val it = load('''
+			module example-system {
+			  yang-version 1.2;
+			  namespace "urn:example:system";
+			  prefix "sys";
+			}
+		''')
+		assertError(root.subStatements.filter(YangVersion).head, INCORRECT_VERSION, 39, 3);
+	}
+
+	@Test
+	def void checkModule_Version_Missing() {
+		val it = load('''
+			module example-system {
+			  namespace "urn:example:system";
+			  prefix "sys";
+			}
+		''');
+		assertError(root, SUBSTATEMENT_CARDINALITY, 7, 14);
+	}
+
+	@Test
+	def void checkModule_Version() {
+		load('''
+			module example-system {
+			  yang-version 1.1;
+			  namespace "urn:example:system";
+			  prefix "sys";
+			}
+		''').assertNoErrors;
+	}
+
+	@Test
+	def void checkModule_Version_Duplicate() {
+		val it = load('''
+			module example-system {
+			  yang-version 1.1;
+			  yang-version 1.1;
+			  namespace "urn:example:system";
+			  prefix "sys";
+			}
+		''');
+		assertError(root, SUBSTATEMENT_CARDINALITY, 26, 17);
+		assertError(root, SUBSTATEMENT_CARDINALITY, 46, 17);
+	}
+
+	@Test
+	def void checkModule_Namespace_Missing() {
+		val it = load('''
+			module example-system {
+			  yang-version 1.1;
+			  prefix "sys";
+			}
+		''');
+		assertError(root, SUBSTATEMENT_CARDINALITY, 7, 14);
+	}
+
+	@Test
+	def void checkModule_Namespace() {
+		load('''
+			module example-system {
+			  yang-version 1.1;
+			  namespace "urn:example:system";
+			  prefix "sys";
+			}
+		''').assertNoErrors;
+	}
+
+	@Test
+	def void checkModule_Namespace_Duplicate() {
+		val it = load('''
+			module example-system {
+			  yang-version 1.1;
+			  namespace "urn:example:system";
+			  namespace "urn:example:system";
+			  prefix "sys";
+			}
+		''');
+		assertError(root, SUBSTATEMENT_CARDINALITY, 46, 31);
+		assertError(root, SUBSTATEMENT_CARDINALITY, 80, 31);
+	}
+
+	@Test
+	def void checkModule_Prefix_Missing() {
+		val it = load('''
+			module example-system {
+			  yang-version 1.1;
+			  namespace "urn:example:system";
+			}
+		''');
+		assertError(root, SUBSTATEMENT_CARDINALITY, 7, 14);
+	}
+
+	@Test
+	def void checkModule_Prefix() {
+		load('''
+			module example-system {
+			  yang-version 1.1;
+			  namespace "urn:example:system";
+			  prefix "sys";
+			}
+		''').assertNoErrors;
+	}
+
+	@Test
+	def void checkModule_Prefix_Order() {
+		val it = load('''
+			module example-system {
+			  namespace "urn:example:system";
+			  yang-version 1.1;
+			  prefix "sys";
+			}
+		''');
+		assertError(root, SUBSTATEMENT_ORDERING, 60, 17);
+	}
+
+	@Test
+	def void checkModule_Prefix_Duplicate() {
```

Comment:
We should not have too many redundant tests at this stage. Since `yang-version`, `namespace` and `prefix` share the same cardinality, it's enough to test violations for one.